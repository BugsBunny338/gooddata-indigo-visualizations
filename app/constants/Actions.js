export const BOOTSTRAP = 'BOOTSTRAP';
export const BOOTSTRAP_DATA = 'BOOTSTRAP_DATA';
export const BOOTSTRAP_ERROR = 'BOOTSTRAP_ERROR';

export const LOGOUT = 'LOGOUT';
export const LOGOUT_DATA = 'LOGOUT_DATA';
export const LOGOUT_ERROR = 'LOGOUT_ERROR';

export const CATALOGUE_UPDATE = 'CATALOGUE_UPDATE';
export const CATALOGUE_UPDATED = 'CATALOGUE_UPDATED';
export const CATALOGUE_SET_ACTIVE_FILTER_INDEX = 'CATALOGUE_SET_ACTIVE_FILTER_INDEX';
export const CATALOGUE_SET_QUERY = 'CATALOGUE_SET_QUERY';
export const CATALOGUE_SET_ACTIVE_DATASET_ID = 'CATALOGUE_SET_ACTIVE_DATASET_ID';

export const DATE_DIMENSIONS_DATA = 'DATE_DIMENSIONS_DATA';

export const DATASETS_REQUEST = 'DATASETS_REQUEST';
export const DATASETS_DATA = 'DATASETS_DATA';

export const ITEM_DETAIL_REQUEST = 'ITEM_DETAIL_REQUEST';
export const ITEM_DETAIL_DATA = 'ITEM_DETAIL_DATA';

export const BUCKETS_SELECT_VISUALIZATION_TYPE = 'BUCKETS_SELECT_VISUALIZATION_TYPE';
export const BUCKETS_SET_BUCKET_ITEM_COLLAPSED = 'BUCKETS_SET_BUCKET_ITEM_COLLAPSED';
export const BUCKETS_SET_BUCKET_ITEM_SHOW_IN_PERCENT = 'BUCKETS_SET_BUCKET_ITEM_SHOW_IN_PERCENT';
export const BUCKETS_SET_BUCKET_ITEM_SHOW_POP = 'BUCKETS_SET_BUCKET_ITEM_SHOW_POP';
export const BUCKETS_SET_BUCKET_ITEM_ADD_FILTER = 'BUCKETS_SET_BUCKET_ITEM_ADD_FILTER';
export const BUCKETS_SET_BUCKET_ITEM_REMOVE_FILTER = 'BUCKETS_SET_BUCKET_ITEM_REMOVE_FILTER';
export const BUCKETS_SET_BUCKET_ITEM_AGGREGATION = 'BUCKETS_SET_BUCKET_ITEM_AGGREGATION';
export const BUCKETS_SET_BUCKET_ITEM_DIMENSION = 'BUCKETS_SET_BUCKET_ITEM_DIMENSION';
export const BUCKETS_SET_BUCKET_ITEM_GRANULARITY = 'BUCKETS_SET_BUCKET_ITEM_GRANULARITY';
