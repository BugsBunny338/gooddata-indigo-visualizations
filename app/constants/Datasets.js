export const ALL_DATA_TRANSLATIONS_KEY = 'datasets.production_data';
export const ALL_DATA_ID = null;
