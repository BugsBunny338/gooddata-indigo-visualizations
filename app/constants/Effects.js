export const REDIRECTION = 'REDIRECTION';
export const SET_HASH = 'SET_HASH';
