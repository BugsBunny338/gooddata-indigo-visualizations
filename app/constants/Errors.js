export const LOAD_ERROR = 'LOAD_ERROR';
export const NOT_AUTHORIZED_ERROR = 'NOT_AUTHORIZED_ERROR';
