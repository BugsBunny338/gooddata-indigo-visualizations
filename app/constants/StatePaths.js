export const BOOTSTRAP = ['bootstrapData'];

export const USER_FIRST_NAME = ['bootstrapData', 'accountSetting', 'firstName'];
export const USER_LAST_NAME = ['bootstrapData', 'accountSetting', 'lastName'];
export const PROFILE_SETTING_SEPARATORS = ['bootstrapData', 'profileSetting', 'separators'];
export const USER_EMAIL = ['bootstrapData', 'accountSetting', 'email'];
export const USER_URI = ['bootstrapData', 'accountSetting', 'links', 'self'];
export const USER_LOGIN_MD5 = ['bootstrapData', 'accountSetting', 'loginMD5'];
export const BOOTSTRAP_DATA_PERMISSIONS = ['bootstrapData', 'permissions'];

export const Permissions = {
    CAN_ACCESS_WORKBENCH: 'canAccessWorkbench',
    CAN_MANAGE_PROJECT_DASHBOARD: 'canManageProjectDashboard'
};

export const PROJECT_TEMPLATE = ['bootstrapData', 'project', 'template'];
export const PROJECT_TITLE = ['bootstrapData', 'project', 'title'];
export const PROJECT_URI = ['bootstrapData', 'project', 'uri'];

export const BRANDING = ['bootstrapData', 'branding'];
export const APPLE_TOUCH_ICON_URL = [...BRANDING, 'appleTouchIconUrl'];
export const IS_BRANDED = [...BRANDING, 'isBranded'];
export const ORGANIZATION_NAME = [...BRANDING, 'organizationName'];
export const FAVICON_URL = [...BRANDING, 'faviconUrl'];
export const APPLICATION_TITLE = [...BRANDING, 'applicationTitle'];

export const MENU_ITEMS = ['header', 'menuItems'];
export const ACCOUNT_MENU_ITEMS = ['header', 'accountMenuItems'];

export const PAGE_TITLE = ['pageTitle'];

export const DEVICE_VIEWPORT = ['meta', 'device', 'viewport'];
export const DEVICE_PIXEL_RATIO = ['meta', 'device', 'pixelRatio'];
export const IS_MOBILE_DEVICE = ['meta', 'device', 'isMobile'];
