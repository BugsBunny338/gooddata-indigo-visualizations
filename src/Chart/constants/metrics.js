export const DEFAULT_FORMAT = '#,##0.00';
