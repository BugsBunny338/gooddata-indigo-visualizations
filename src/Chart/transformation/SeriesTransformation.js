import { compact, uniqBy, map, times, constant, set, sortBy, get } from 'lodash';
import { enableDrillablePoints } from '../../utils/drilldownEventing';

// get unique elements for given index regarding their interal id
function getElements(data, index) {
    return uniqBy(map(data.rawData, index), 'id');
}

// get series data
function getSeries(data, seriesNames, categories, indices, drillableItems = []) {
    const seriesData = seriesNames.reduce((acc, series) => {
        return set(acc, series.id, times(categories.length, constant(null)));
    }, {});

    // prepare data points
    if (indices.series !== -1) {
        if (indices.category !== -1) {
            const categoriesIndex = categories.reduce((acc, category, idx) => {
                return category ? set(acc, category.id, idx) : acc;
            }, {});

            data.rawData.forEach((dataRow) => {
                const categoryItemIndex = categoriesIndex[dataRow[indices.category].id];
                const pointDrill = enableDrillablePoints(
                    drillableItems,
                    dataRow[indices.metric],
                    dataRow.slice(0, -1)
                );
                set(seriesData, [dataRow[indices.series].id, categoryItemIndex], pointDrill);
            });
        } else {
            data.rawData.forEach((dataRow) => {
                const pointDrill = enableDrillablePoints(
                    drillableItems,
                    dataRow[indices.metric],
                    dataRow.slice(0, -1)
                );
                set(seriesData, [dataRow[indices.series].id, 0], pointDrill);
            });
        }
    }

    // return series in the same order as requested in seriesNames
    return seriesNames.map((name, index) => {
        return {
            name: name.value,
            legendIndex: index,
            data: seriesData[name.id]
        };
    });
}

function setDrillableFlag(series) {
    return series.map((s) => {
        return {
            ...s,
            isDrillable: get(s, 'data.0.drilldown')
        };
    });
}

/**
 * Get highcharts-compatible charting data from the data which
 * already passed a metricNames+metricValues transformation
 * Needs configuration for categories/series and metric values
 *
 * Input
 * Example:
 * <pre>
 *
 *   Input
 *
 *     A  metricNames metricValues
 *    ------------------------------
 *     a       M1         1111    // first metric
 *     b       M1         1112
 *     c       M1         1121
 *     a       M2         2333    // second metric
 *     b       M2         2334
 *     c       M2         2345
 *
 *   Input configuration:
 *      sort: false,
 *      indices: {
 *          categories: 0,
 *          series: 1,
 *          metric: 2
 *      }
 *
 * </pre>
 *
 * Output
 * Example:
 *      {
 *          categories: ['a', 'b', 'c'],
 *          series: [
 *              {
 *                  name: 'M1',
 *                  data: [ { y: 1111 }, { y: 1112 }, { y: 1121 } ]
 *              },
 *              {
 *                  name: 'M2',
 *                  data: [ { y: 2333 }, { y: 2334 }, { y: 2345 } ]
 *              }
 *          ]
 *      }
 * <pre>
 *
 * </pre>
 **/
export function getChartData(data, configuration, drillableItems) {
    const indices = configuration.indices;

    const categories = getElements(data, indices.category);
    const seriesNames = getElements(data, indices.series);

    const seriesNamesSorted = configuration.sortSeries ? sortBy(seriesNames, 'value') : seriesNames;
    const seriesNamesCompact = compact(seriesNamesSorted);

    const series = getSeries(data, seriesNamesCompact, categories, indices, drillableItems);

    return {
        categories: map(categories, 'value'),
        series: setDrillableFlag(series)
    };
}
