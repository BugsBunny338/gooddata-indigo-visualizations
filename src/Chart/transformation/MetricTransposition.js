import { cloneDeep } from 'lodash';
import { enrichHeaders } from './EnrichHeaders';
import { parseValue } from '../../utils/common';

import { DEFAULT_FORMAT } from '../constants/metrics';

function invalidInput(data) {
    return !data || data.isLoading;
}

function transposeData(headers, metrics, rawData) {
    const data = [];

    for (let mi = 0; mi < metrics.length; mi++) {
        for (let ri = 0; ri < rawData.length; ri++) {
            const row = [];

            for (let ci = 0; ci < headers.length; ci++) {
                if (headers[ci].type === 'attrLabel') {
                    const element = rawData[ri][ci];

                    row.push({
                        id: element.id,
                        value: element.name,
                        header: {
                            id: headers[ci].id,
                            uri: headers[ci].uri
                        }
                    });
                }
            }

            const metric = metrics[mi];

            row.push({
                id: metric.header.id, // each metric must have unique id
                value: metric.header.title,
                header: {
                    id: metric.header.id ? metric.header.id : null,
                    uri: metric.header.uri ? metric.header.uri : null
                }
            });

            const metricValue = rawData[ri][metric.index];
            const y = parseValue(metricValue);
            row.push({
                format: metric.header.format || DEFAULT_FORMAT,
                y,
                marker: { enabled: y !== null }
            });

            data.push(row);
        }
    }

    return data;
}

export function transposeMetrics(data) {
    if (invalidInput(data)) {
        return data;
    }

    const dataCopy = cloneDeep(data);

    const { headers, metrics } = enrichHeaders(dataCopy.headers);

    if (metrics.length < 1) return dataCopy;

    return {
        ...dataCopy,
        headers,
        rawData: transposeData(dataCopy.headers, metrics, dataCopy.rawData)
    };
}
