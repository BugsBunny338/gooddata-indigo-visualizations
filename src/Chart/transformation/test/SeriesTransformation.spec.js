import { map } from 'lodash';
import { DEFAULT_FORMAT } from '../../constants/metrics';
import { getChartData } from '../SeriesTransformation';
import { transposeMetrics } from '../MetricTransposition';
import { singleMeasure, singleMeasureAndAttribute } from './fixtures/executionResponses';

describe('Getting chart data', () => {
    it('should create series from single measure', () => {
        const data = transposeMetrics(singleMeasure);

        const indices = {
            category: -1,
            series: 0,
            metric: 1
        };

        const configuration = {
            indices
        };

        expect(getChartData(data, configuration)).toMatchSnapshot();
    });

    it('should create series from single measure and attribute', () => {
        const data = transposeMetrics(singleMeasureAndAttribute);

        const indices = {
            category: 0,
            series: 1,
            metric: 2
        };

        const configuration = {
            indices
        };

        expect(getChartData(data, configuration)).toMatchSnapshot();
    });

    it('should correctly transform with metrics with/without formats', () => {
        const data = {
            headers: [
                { id: 'a1', type: 'attrLabel' },
                {
                    id: 'metricNames',
                    type: 'attrLabel',
                    metrics: {
                        M1: { header: { format: '#.## M1' } },
                        M2: { header: { format: '#.## M2' } },
                        M3: { header: { } }
                    }
                },
                { id: 'metricValues', type: 'metric' }
            ],
            /* eslint-disable max-len */
            rawData: [
                [{ id: 'a', value: 'a' }, { id: 'M1', value: 'M1' }, { y: 1, format: '#.## M1' }],
                [{ id: 'b', value: 'b' }, { id: 'M1', value: 'M1' }, { y: 2, format: '#.## M1' }],
                [{ id: 'c', value: 'c' }, { id: 'M1', value: 'M1' }, { y: 3, format: '#.## M1' }],
                [{ id: 'a', value: 'a' }, { id: 'M2', value: 'M2' }, { y: 4, format: '#.## M2' }],
                [{ id: 'b', value: 'b' }, { id: 'M2', value: 'M2' }, { y: 5, format: '#.## M2' }],
                [{ id: 'c', value: 'c' }, { id: 'M2', value: 'M2' }, { y: 6, format: '#.## M2' }],
                [{ id: 'a', value: 'a' }, { id: 'M3', value: 'M3' }, { y: 7, format: DEFAULT_FORMAT }],
                [{ id: 'b', value: 'b' }, { id: 'M3', value: 'M3' }, { y: 8, format: DEFAULT_FORMAT }],
                [{ id: 'c', value: 'c' }, { id: 'M3', value: 'M3' }, { y: 9, format: DEFAULT_FORMAT }]
            ]
            /* eslint-enable max-len */
        };

        const configuration = {
            indices: {
                category: 0,
                series: 1,
                metric: 2
            },
            sortSeries: true
        };

        const expectedData = {
            categories: ['a', 'b', 'c'],
            series: [
                {
                    name: 'M1',
                    data: [
                        { y: 1, format: '#.## M1' },
                        { y: 2, format: '#.## M1' },
                        { y: 3, format: '#.## M1' }
                    ],
                    legendIndex: 0
                },
                {
                    name: 'M2',
                    data: [
                        { y: 4, format: '#.## M2' },
                        { y: 5, format: '#.## M2' },
                        { y: 6, format: '#.## M2' }
                    ],
                    legendIndex: 1
                },
                {
                    name: 'M3',
                    data: [
                        { y: 7, format: DEFAULT_FORMAT },
                        { y: 8, format: DEFAULT_FORMAT },
                        { y: 9, format: DEFAULT_FORMAT }
                    ],
                    legendIndex: 2
                }

            ]
        };

        expect(getChartData(data, configuration)).toEqual(expectedData);
    });

    it('should correctly transform when multiple attributes present', () => {
        const data = {
            headers: [
                { id: 'a1', type: 'attrLabel' },
                { id: 'a2', type: 'attrLabel' },
                {
                    id: 'metricNames',
                    type: 'attrLabel',
                    metrics: {
                        M1: { header: { format: '#.## M1' } }
                    }
                },
                { id: 'metricValues', type: 'metric' }
            ],
            /* eslint-disable max-len */
            rawData: [
                [{ id: 'a', value: 'a' }, { id: 'x', value: 'x' }, { id: 'M1', value: 'M1' }, { y: 1, format: '#.## M1' }],
                [{ id: 'a', value: 'a' }, { id: 'y', value: 'y' }, { id: 'M1', value: 'M1' }, { y: 2, format: '#.## M1' }],
                [{ id: 'b', value: 'b' }, { id: 'x', value: 'x' }, { id: 'M1', value: 'M1' }, { y: 3, format: '#.## M1' }],
                [{ id: 'b', value: 'b' }, { id: 'y', value: 'y' }, { id: 'M1', value: 'M1' }, { y: 4, format: '#.## M1' }],
                [{ id: 'c', value: 'c' }, { id: 'x', value: 'x' }, { id: 'M1', value: 'M1' }, { y: 5, format: '#.## M1' }],
                [{ id: 'c', value: 'c' }, { id: 'y', value: 'y' }, { id: 'M1', value: 'M1' }, { y: 6, format: '#.## M1' }]
            ]
            /* eslint-enable max-len */
        };

        const configuration = {
            indices: {
                category: 0,
                series: 1,
                metric: 3
            },
            sortSeries: true
        };

        const expectedData = {
            categories: ['a', 'b', 'c'],
            series: [
                {
                    name: 'x',
                    data: [
                        { y: 1, format: '#.## M1' },
                        { y: 3, format: '#.## M1' },
                        { y: 5, format: '#.## M1' }
                    ],
                    legendIndex: 0
                },
                {
                    name: 'y',
                    data: [
                        { y: 2, format: '#.## M1' },
                        { y: 4, format: '#.## M1' },
                        { y: 6, format: '#.## M1' }
                    ],
                    legendIndex: 1
                }
            ]
        };

        expect(getChartData(data, configuration)).toEqual(expectedData);
    });

    it('should produce sorted legend even if some values are missing', () => {
        const data = {
            headers: [
                { id: 'a1', type: 'attrLabel' },
                { id: 'a2', type: 'attrLabel' },
                {
                    id: 'metricNames',
                    type: 'attrLabel',
                    metrics: {
                        M1: { header: { format: '#.## M1' } }
                    }
                },
                { id: 'metricValues', type: 'metric' }
            ],
            /* eslint-disable max-len */
            rawData: [
                [{ id: 'a', value: 'a' }, { id: 'x', value: 'x' }, { id: 'M1', value: 'M1' }, { y: 1, format: '#.## M1' }],
                [{ id: 'a', value: 'a' }, { id: 'z', value: 'z' }, { id: 'M1', value: 'M1' }, { y: 2, format: '#.## M1' }],
                [{ id: 'b', value: 'b' }, { id: 'x', value: 'x' }, { id: 'M1', value: 'M1' }, { y: 3, format: '#.## M1' }],
                [{ id: 'b', value: 'b' }, { id: 'y', value: 'y' }, { id: 'M1', value: 'M1' }, { y: 4, format: '#.## M1' }],
                [{ id: 'b', value: 'b' }, { id: 'z', value: 'z' }, { id: 'M1', value: 'M1' }, { y: 5, format: '#.## M1' }]
            ]
            /* eslint-enable max-len */
        };

        const configuration = {
            indices: {
                category: 0,
                series: 1,
                metric: 3
            },
            sortSeries: true
        };

        const expectedSeries = ['x', 'y', 'z'];
        const series = map(getChartData(data, configuration).series, 'name');

        expect(series).toEqual(expectedSeries);
    });
});
