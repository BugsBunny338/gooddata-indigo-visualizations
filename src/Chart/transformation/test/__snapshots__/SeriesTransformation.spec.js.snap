// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Getting chart data should create series from single measure 1`] = `
Object {
  "categories": Array [
    undefined,
  ],
  "series": Array [
    Object {
      "data": Array [
        Object {
          "format": "#,##0",
          "marker": Object {
            "enabled": true,
          },
          "y": 118,
        },
      ],
      "legendIndex": 0,
      "name": "# of Accounts",
    },
  ],
}
`;

exports[`Getting chart data should create series from single measure and attribute 1`] = `
Object {
  "categories": Array [
    "",
    "Churned",
    "N/A",
    "OK",
    "Potentially Churned / Salesforce Issues",
  ],
  "series": Array [
    Object {
      "data": Array [
        Object {
          "format": "#,##0",
          "marker": Object {
            "enabled": true,
          },
          "y": 1,
        },
        Object {
          "format": "#,##0",
          "marker": Object {
            "enabled": true,
          },
          "y": 25,
        },
        Object {
          "format": "#,##0",
          "marker": Object {
            "enabled": true,
          },
          "y": 1,
        },
        Object {
          "format": "#,##0",
          "marker": Object {
            "enabled": true,
          },
          "y": 153,
        },
        Object {
          "format": "#,##0",
          "marker": Object {
            "enabled": true,
          },
          "y": 27,
        },
      ],
      "legendIndex": 0,
      "name": "# of Accounts",
    },
  ],
}
`;
